plugins/experimental/server_push_preload/server_push_preload.cc

```diff
@@ -0,0 +1,109 @@
+#include <vector>
+#include <ts/ts.h>
+#include <ts/experimental.h>
+#include <atscppapi/GlobalPlugin.h>
+#include <atscppapi/utils.h>
+
+#define PLUGIN_NAME "server_push_preload"
+
+using namespace std;
+using namespace atscppapi;
+
+static regex linkRegexp("<([^>]+)>;.*rel=preload");
+
+namespace
+{
+GlobalPlugin *plugin;
+}
+
+class LinkServerPushPlugin : public GlobalPlugin
+{
+public:
+  LinkServerPushPlugin()
+  {
+    TSDebug(PLUGIN_NAME, "registering transaction hooks");
+    LinkServerPushPlugin::registerHook(HOOK_SEND_RESPONSE_HEADERS);
+  }
+
+  virtual void
+  handleSendResponseHeaders(Transaction &transaction)
+  {
+    serverPush(transaction);
+    transaction.resume();
+  }
+
+  void
+  serverPush(Transaction &transaction)
+  {
+    ClientRequest &request = transaction.getClientRequest();
+    if (request.getMethod() != HTTP_METHOD_GET) {
+      return;
+    }
+
+    Response &response = transaction.getClientResponse();
+    Headers &headers = response.getHeaders();
+
+    TSHttpTxn txnp = static_cast<TSHttpTxn>(transaction.getAtsHandle());
+    string linkHeader = headers.values("Link");
+    const Url &clientUrl = request.getPristineUrl();
+
+    if (linkHeader.empty()) {
+      TSDebug(PLUGIN_NAME, "Link header not found for url: %s", clientUrl.getUrlString().c_str());
+      return;
+    }
+
+    TSDebug(PLUGIN_NAME, "Found Link header: %s", linkHeader.c_str());
+    stringstream ss(linkHeader);
+    vector<string> links = split(linkHeader, ',');
+
+    for (const string& link : links) {
+      TSDebug(PLUGIN_NAME, "Link: %s", link.c_str());
+      if (link.find("nopush") != string::npos) {
+        TSDebug(PLUGIN_NAME, "Ignoring nopush link: %s", link.c_str());
+        continue;
+      }
+
+      TSDebug(PLUGIN_NAME, "Parsing link header: %s", link.c_str());
+      smatch matches;
+      if(regex_search(link, matches, linkRegexp)) {
+        string url = matches[1].str();
+        TSDebug(PLUGIN_NAME, "Found link header match: %s", url.c_str());
+
+        Request request(url);
+        Url &linkUrl = request.getUrl();
+
+        if (linkUrl.getHost().empty()) {
+          linkUrl.setHost(clientUrl.getHost());
+          linkUrl.setScheme(clientUrl.getScheme());
+        }
+        string lu = linkUrl.getUrlString();
+        TSDebug(PLUGIN_NAME, "Push preloaded content: %s", lu.c_str());
+        TSHttpTxnServerPush(txnp, lu.c_str(), lu.length());
+      } else {
+        TSDebug(PLUGIN_NAME, "No match found for link header: %s", link.c_str());
+      }
+    }
+  }
+
+  vector<string>
+  split(const string &s, char delim) {
+    stringstream ss(s);
+    string item;
+    vector<string> tokens;
+    while (getline(ss, item, delim)) {
+      tokens.push_back(item);
+    }
+    return tokens;
+  }
+};
+
+void
+TSPluginInit(int argc ATSCPPAPI_UNUSED, const char *argv[] ATSCPPAPI_UNUSED)
+{
+  TSDebug(PLUGIN_NAME, "Init");
+  RegisterGlobalPlugin("ServerPushPreloadPlugin", PLUGIN_NAME, "david.calavera@gmail.com");
```

Comment:
I'm not 100% sure but I think we should use dev@trafficserver.apache.com here and you can claim copyright in NOTICE file if you want.
@zwoop Am I correct?